plugin/pkg/replacer/replacer.go

```diff
@@ -36,80 +41,92 @@ type replacer struct {
+		ctx: ctx,
+		replacements: map[string]Func{
+			"{type}":  func() string { return req.Type() },
+			"{name}":  func() string { return req.Name() },
+			"{class}": func() string { return req.Class() },
+			"{proto}": func() string { return req.Proto() },
 			"{when}": func() string {
 				return time.Now().Format(timeFormat)
-			}(),
-			"{size}":   strconv.Itoa(req.Len()),
-			"{remote}": addrToRFC3986(req.IP()),
-			"{port}":   req.Port(),
-			"{local}":  addrToRFC3986(req.LocalIP()),
-			"{forward/resolving_proxy}": getMetadata(ctx, "forward/resolving_proxy"),
+			},
+			"{size}":   func() string { return strconv.Itoa(req.Len()) },
+			"{remote}": func() string { return addrToRFC3986(req.IP()) },
+			"{port}":   func() string { return req.Port() },
+			"{local}":  func() string { return addrToRFC3986(req.LocalIP()) },
 		},
+		evaluated:  make(map[string]string, 0),
 		emptyValue: emptyValue,
 	}
 	if rr != nil {
 		rcode := dns.RcodeToString[rr.Rcode]
 		if rcode == "" {
 			rcode = strconv.Itoa(rr.Rcode)
 		}
-		rep.replacements["{rcode}"] = rcode
-		rep.replacements["{rsize}"] = strconv.Itoa(rr.Len)
-		rep.replacements["{duration}"] = strconv.FormatFloat(time.Since(rr.Start).Seconds(), 'f', -1, 64) + "s"
+		rep.replacements["{rcode}"] = func() string { return rcode }
+		rep.replacements["{rsize}"] = func() string { return strconv.Itoa(rr.Len) }
+		rep.replacements["{duration}"] = func() string { return strconv.FormatFloat(time.Since(rr.Start).Seconds(), 'f', -1, 64) + "s" }
 		if rr.Msg != nil {
-			rep.replacements[headerReplacer+"rflags}"] = flagsToString(rr.Msg.MsgHdr)
-			rep.replacements["{A}"] = answersCount(rr.Msg, dns.TypeA)
-			rep.replacements["{AAAA}"] = answersCount(rr.Msg, dns.TypeAAAA)
+			rep.replacements[headerReplacer+"rflags}"] = func() string { return flagsToString(rr.Msg.MsgHdr) }
 		}
 	}
 
 	// Header placeholders (case-insensitive)
-	rep.replacements[headerReplacer+"id}"] = strconv.Itoa(int(r.Id))
-	rep.replacements[headerReplacer+"opcode}"] = strconv.Itoa(r.Opcode)
-	rep.replacements[headerReplacer+"do}"] = boolToString(req.Do())
-	rep.replacements[headerReplacer+"bufsize}"] = strconv.Itoa(req.Size())
+	rep.replacements[headerReplacer+"id}"] = func() string { return strconv.Itoa(int(r.Id)) }
+	rep.replacements[headerReplacer+"opcode}"] = func() string { return strconv.Itoa(r.Opcode) }
+	rep.replacements[headerReplacer+"do}"] = func() string { return boolToString(req.Do()) }
+	rep.replacements[headerReplacer+"bufsize}"] = func() string { return strconv.Itoa(req.Size()) }
 
 	return rep
 }
 
 // Replace performs a replacement of values on s and returns
 // the string with the replaced values.
 func (r replacer) Replace(s string) string {
-	// Header replacements - these are case-insensitive, so we can't just use strings.Replace()
-	for strings.Contains(s, headerReplacer) {
-		idxStart := strings.Index(s, headerReplacer)
-		endOffset := idxStart + len(headerReplacer)
+	b := strings.Builder{}
```

Comment:
this is a good fix, because looking through this replacement code I think it looks pretty heavy weight

Reply:
ideally, single strings.Builder{} object should be used for building whole replacement string